core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1167,19 +1122,9 @@ class KafkaController(val config: KafkaConfig, zkUtils: ZkUtils, val brokerState
      * @throws Exception On any error.
      */
     @throws[Exception]
-    def handleNewSession() {
-      info("ZK expired; shut down all controller components and try to re-elect")
-      if (controllerElector.getControllerID() != config.brokerId) {
-        onControllerResignation()
-        inLock(controllerContext.controllerLock) {
-          controllerElector.elect
-        }
-      } else {
-        // This can happen when there are multiple consecutive session expiration and handleNewSession() are called multiple
-        // times. The first call may already register the controller path using the newest ZK session. Therefore, the
-        // controller path will exist in subsequent calls to handleNewSession().
-        info("ZK expired, but the current controller id %d is the same as this broker id, skip re-elect".format(config.brokerId))
-      }
+    def handleNewSession(): Unit = {
+      controllerEventQueue.put(Resign(getControllerID()))
+      controllerEventQueue.put(Elect)
```

Comment:
That's an interesting point.

In the current PR, only the ZkClient's ZkEventThread enqueues Resign and Elect. This happens in ControllerChangeListener.handleDataDeleted and SessionExpirationListener.handleNewSession. Since ZkEventThread can't interleave itself, this leaves us with only two types of threads that can interleave an event between a Resign and Elect:
1. the scheduler can interleave an UpdateMetrics, TopicDeletionProgressCheck, or AutoPreferredReplicaLeaderElection.
2. the RequestSendThreads can interleave a TopicDeletionStopReplicaResult from the StopReplicaRequest callback.

I'm not too worried about missing one of the above events interleaved between resignation and election since an elected controller should be able to pick up where things left off by reading zookeeper state.

I am worried however about processing interleaved events while not the controller. For instance, I just noticed a bug in the PR where TopicDeletionProgressCheck and TopicDeletionStopReplicaResult don't first check isActive. I'll update the PR now.